execute/plugin.go

```diff
@@ -199,6 +216,248 @@ func (p *Plugin) ObservationQuorum(outctx ocr3types.OutcomeContext, query types.
+		if searchErr != nil {
+			searchErr = fmt.Errorf("unable to build a single chain report (messages %d): %w", mid, err)
+		}
+
+		if (encodedSize2) <= maxSizeBytes {
+			// mid is a valid report size, try something bigger next iteration.
+			finalReport = finalReport2
+			encodedSize = encodedSize2
+			return false // not full
+		}
+		return true // full
+	})
+	if searchErr != nil {
+		return cciptypes.ExecutePluginReportSingleChain{}, 0, cciptypes.ExecutePluginCommitDataWithMessages{}, searchErr
+	}
+
+	// No messages fit into the report.
+	if idx <= 0 {
+		return cciptypes.ExecutePluginReportSingleChain{},
+			0,
+			cciptypes.ExecutePluginCommitDataWithMessages{},
+			errNothingExecuted
+	}
+
+	report = markNewMessagesExecuted(finalReport, report)
+	return finalReport, encodedSize, report, nil
+}
+
+// buildSingleChainReport converts the on-chain event data stored in cciptypes.ExecutePluginCommitDataWithMessages into
+// the final on-chain report format.
+//
+// The hasher and encoding codec are provided as arguments to allow for chain-specific formats to be used.
+//
+// The maxMessages argument is used to limit the number of messages that are included in the report. If maxMessages is
+// set to 0, all messages will be included. This allows the caller to create smaller reports if needed.
+func buildSingleChainReport(
+	ctx context.Context,
+	lggr logger.Logger,
+	hasher cciptypes.MessageHasher,
+	tokenDataReader TokenDataReader,
+	encoder cciptypes.ExecutePluginCodec,
+	report cciptypes.ExecutePluginCommitDataWithMessages,
+	maxMessages int,
+) (cciptypes.ExecutePluginReportSingleChain, int, error) {
+	// TODO: maxMessages selects messages in FIFO order which may not yield the optimal message size. One message with a
+	//       maximum data size could push the report over a size limit even if several smaller messages could have fit.
+	if maxMessages == 0 {
+		maxMessages = len(report.Messages)
+	}
+
+	tree, err := constructMerkleTree(ctx, hasher, report)
+	if err != nil {
+		return cciptypes.ExecutePluginReportSingleChain{}, 0,
+			fmt.Errorf("unable to construct merkle tree from messages: %w", err)
+	}
+	lggr.Debugw(
+		"constructing merkle tree",
+		"sourceChain", report.SourceChain,
+		"treeLeaves", len(report.Messages))
+	numMsgs := len(report.Messages)
+
+	// Iterate sequence range and executed messages to select messages to execute.
+	var toExecute []int
+	var offchainTokenData [][][]byte
+	var msgInRoot []cciptypes.CCIPMsg
+	executedIdx := 0
+	for i := 0; i < numMsgs && len(toExecute) <= maxMessages; i++ {
+		seqNum := report.SequenceNumberRange.Start() + cciptypes.SeqNum(i)
+		// Skip messages which are already executed
+		if executedIdx < len(report.ExecutedMessages) && report.ExecutedMessages[executedIdx] == seqNum {
+			executedIdx++
+		} else {
+			msg := report.Messages[i]
+			tokenData, err := tokenDataReader.ReadTokenData(context.Background(), report.SourceChain, msg.SeqNum)
```

Comment:
In addition to the readiness of the token data, there are things like e.g gas limit that determine whether we can include a message in a batch. Do we have tickets for that?

Reply:
Thanks, that one wasn't on my radar. [Ticket created.](https://smartcontract-it.atlassian.net/browse/CCIP-2684)